[WebGPU] GPUCommandEncoder.clearBuffer does not appear to work

mwyrzykowski opened the pull request:
#### 033f8d13532716f63ffa02f477d541e9613c4eba
<pre>
[WebGPU] GPUCommandEncoder.clearBuffer does not appear to work
<a href="https://bugs.webkit.org/show_bug.cgi?id=252643">https://bugs.webkit.org/show_bug.cgi?id=252643</a>
&lt;radar://105708271&gt;

Reviewed by Myles C. Maxfield.

Changes were not being reflected because useResoure was not
being called.

* Source/WebCore/Modules/WebGPU/GPUCommandEncoder.cpp:
(WebCore::GPUCommandEncoder::clearBuffer):
* Source/WebCore/Modules/WebGPU/GPUCommandEncoder.h:
* Source/WebCore/Modules/WebGPU/GPUCommandEncoder.idl:
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::clearBuffer):

Canonical link: <a href="https://commits.webkit.org/260707@main">https://commits.webkit.org/260707@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66d3f8a7ffdaca901c5183ba66bcd7c0aea0f702

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109015 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18094 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41836 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/550 "Updated wpe dependencies (failure)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118291 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19550 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9394 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101260 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114773 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14659 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97895 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42801 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96631 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29537 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84549 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10879 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30885 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11623 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7820 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17003 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50483 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7390 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13222 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -486,6 +486,38 @@ static bool validateWriteTexture(const WGPUImageCopyTexture& destination, const
+    for (id<MTLFence> fence in pendingFences)
+        callback(fence);
+
+    [pendingFences removeAllObjects];
```
This assumes that fences added to one command encoder are all used in the command encoder processed next. Could there ever be a situation where WGPU would need to wait for different fences in a compute and render encoder from the same command encoder? If so, I'd like to see a `FIXME` at least.

mwyrzykowski replied:
Since I moved away from fences, I will not worry about this situation for now

djg commented on Source/WebCore/Modules/WebGPU/GPUCommandEncoder.idl:
```diff
@@ -62,9 +62,10 @@ interface GPUCommandEncoder {
     undefined clearBuffer(
         GPUBuffer buffer,
-        optional GPUSize64 offset = 0,
+        optional GPUSize64 offset,
```
Why was this change necessary?

mwyrzykowski replied:
It’s a workaround for the bug mentioned in the FIXME above

djg reviewed (commented):
Looks great. 🚢

litherum commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -126,6 +126,7 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
 {
     if (m_blitCommandEncoder) {
         [m_blitCommandEncoder endEncoding];
+        [m_blitCommandEncoder updateFence:m_device->getQueue().generateFenceToWaitFor()];
```
This shouldn't be necessary.

mwyrzykowski replied:
> We shouldn't need fences to do this; Metal has automatic hazard tracking. It is correct to use the Blit command encoder to fill a buffer with 0s and then just use it in a subsequent command/pass without any fences.

That was my expectation too but it does not seem to work. Let me take a look with the debugger and see what is occurring

litherum commented on Source/WebGPU/WebGPU/ComputePassEncoder.mm:
```diff
@@ -71,11 +71,15 @@
 
 void ComputePassEncoder::dispatch(uint32_t x, uint32_t y, uint32_t z)
 {
+    m_device->getQueue().waitForAndClearPendingFences(m_computeCommandEncoder);
```
I think this is wrong.

mwyrzykowski replied:
Indeed :) I was missing a call to useResource, which is why the hazard tracking wasn't working

litherum reviewed (changes requested):
We shouldn't need fences to do this; Metal has automatic hazard tracking. It is correct to use the Blit command encoder to fill a buffer with 0s and then just use it in a subsequent command/pass without any fences.

mwyrzykowski committed:
[WebGPU] GPUCommandEncoder.clearBuffer does not appear to work
https://bugs.webkit.org/show_bug.cgi?id=252643
<radar://105708271>

Reviewed by Myles C. Maxfield.

Changes were not being reflected because useResoure was not
being called.

* Source/WebCore/Modules/WebGPU/GPUCommandEncoder.cpp:
(WebCore::GPUCommandEncoder::clearBuffer):
* Source/WebCore/Modules/WebGPU/GPUCommandEncoder.h:
* Source/WebCore/Modules/WebGPU/GPUCommandEncoder.idl:
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::clearBuffer):

Canonical link: https://commits.webkit.org/260707@main
